src/index.ts

```diff
@@ -1,58 +1,185 @@
+        const pubkey = this.suite.point().one().mul(secret);
+        return {
+          secret,
+          pubkey,
+        };
+      } catch (e) {
+        n++;
+      }
+    }
+  }
+
+  /**
+   * Convert a public key from EVM into Medusa format
+   * @param {EVMG1Point} pubkey to decode
+   * @returns {P} A Medusa Public Key
+   */
+  decodePublicKey(pubkey: EVMG1Point): P {
+    return this.suite.point().fromEvm(pubkey)._unsafeUnwrap();
+  }
+
+  /**
+   * Encrypt a message for a user; include a label to prevent replay attacks via onchain DLEQ verification
+   * @param {Uint8Array} data to encrypt
+   * @param {P} medusaPublicKey of the encryption oracle to encrypt towards
+   * @param {string} contractAddress of the application developer's contract to be included in the label
+   * @param {string} userAddress of the user to be included in the label
+   * @returns {Promise<{encryptedData: string, encryptedKey: HGamalEVMCipher}>} The encrypted data and the EVM encoded encrypted key
+   */
+  async encrypt(
+    data: Uint8Array,
+    medusaPublicKey: P,
+    contractAddress: `0x${string}`,
+    userAddress: `0x${string}`
+  ): Promise<{ encryptedData: string; encryptedKey: HGamalEVMCipher }> {
+    const hgamalSuite = new HGamalSuite(this.suite);
+    const label = Label.from(medusaPublicKey, contractAddress, userAddress);
+    const bundle = (
+      await hgamalSuite.encryptToMedusa(data, medusaPublicKey, label)
+    )._unsafeUnwrap();
+
+    return {
+      encryptedKey: bundle.encryptedKey.toEvm(),
+      encryptedData: Base64.fromUint8Array(bundle.encryptedData),
+    };
+  }
+
+  /**
+   * Decrypt a message that has been reencrypted for a user
+   * @param {HGamalEVMCipher} ciphertext of encrypted key to decrypt encryptedContents
+   * @param {string} encryptedContents to decrypt
+   * @param {S} userPrivateKey of the user to decrypt ciphertext
+   * @param {P} medusaPublicKey of the medusa encryption oracle
+   * @returns {Promise<string>} The decrypted data
+   */
+  async decrypt(
+    ciphertext: HGamalEVMCipher,
+    encryptedContents: string,
+    userPrivateKey: S,
+    medusaPublicKey: P
+  ): Promise<string> {
+    // Base64 decode into Uint8Array
+    const encryptedData = Base64.toUint8Array(encryptedContents);
+
+    const hgamalSuite = new HGamalSuite(this.suite);
+
+    // Convert the ciphertext to a format that the Medusa SDK can use
+    const cipher = HGamalCipher.default(this.suite)
+      .fromEvm(ciphertext)
+      ._unsafeUnwrap();
+
+    // Create bundle with encrypted data and extraneous cipher (not used)
+    const bundle: EncryptionBundle<HGamalEVMCipher, HGamalCipher<S, P>> = {
+      encryptedData,
+      encryptedKey: cipher,
+    };
+
+    // Decrypt
+    const decryptionRes = await hgamalSuite.decryptFromMedusa(
+      userPrivateKey,
+      medusaPublicKey,
+      bundle,
+      cipher
+    );
+    // Decode to string
+    return new TextDecoder().decode(decryptionRes._unsafeUnwrap());
```

Comment:
DO we really want to do that for the user ? It is not consistent with the encryption method that takes a uint8array. I think we should return a uint8array as well here and let the user decide what it wants to do with it.